### PR TITLE
Feature/hzn 104 tests failing to run on develop 

### LINKED ISSUE #103: [HZN-104] Tests failing to run on develop branch (26/03/2020)
Transferred by @LaurenceJackson on Mar 26, 2021, 13:57

:::Issue recreated from Jira issue [HZN-104](https://gsttmri.atlassian.net/browse/HZN-104):::

Error received below when trying to run ghosting tests on origin/develop branch on 26/03/2020. Also tried to run SNR, same error received. Appears to be related to recent changes to relaxometry code.

(hazen-venv) (base) Janes-MacBook:cavell_2021 Ansell$ hazen ghosting nsa1
Traceback (most recent call last):
  File "/Users/Ansell/hazen-venv/bin/hazen", line 11, in <module>
    load_entry_point('hazen', 'console_scripts', 'hazen')()
  File "/Users/Ansell/hazen_code/hazenlib/__init__.py", line 326, in main
    relaxometry_cli_args_used = [arg for arg in relaxometry_cli_args if arguments[arg]]
  File "/Users/Ansell/hazen_code/hazenlib/__init__.py", line 326, in <listcomp>
    relaxometry_cli_args_used = [arg for arg in relaxometry_cli_args if arguments[arg]]
KeyError: '--verbose'

### PR DESCRIPTION
**Pull request by pcw24601**
Fixes the problem on my Windows machine. However, the unittest spawns a subprocess to test the CLI and I don't know how to adapt to work on other systems.

### CODE REVIEW

**Comment by hshuaib90** [comment]
HI @pcw24601 - is this ready for review?

**Comment by pcw24601** [comment]
@hshuaib90 I'm still a bit stuck with this :-(.

TL;DR: The initial problem is fixed but the test to check it doesn't work.

I believe I've fixed the initial problem (numbers in command line arguments being passed as strings to functions expecting a number). I've also written a test which checks the CLI works by spawning a subprocess. The test passes on my Windows machine. However, it failed the testing pipeline under Bitbucket and you previously mentioned an install step to add `hazen`  to the path was needed. I've no idea how to do this--any pointers? I can delete the test if that's easier.

**Comment by laurencejackson** [comment]
Hey @pcw24601 I'm going to take a look at this too!

**Comment by laurencejackson** [comment]
Hi @pcw24601, I think resolving this will require us to merge in the current github actions since we are no longer using the bitbucket pipelines. 

Rather than write a unit test that will check that the CLI functions correctly using subprocesses, I think it will be simpler and just as effective to run it in the test environment and use the exit code as a marker of pass/fail.

Could you have a look at the current develop branch and copy over the .github folder in the root directory into your root directory? This directory contains 2 workflows, one is for unit testing and the other is for deployment to dockerhub. I think we should add a third which will perform the CLI test.

In the .github/workflows folder could you add a new workflow called something like .github/workflows/cli-test.yml with something like this:

```name: Test CLI

on:
  pull_request:
    branches:
      - '*'

jobs:
  test:
    runs-on: ubuntu-latest

    - name: Set up Python 3.6
      uses: actions/setup-python@v2
      with:
        python-version: 3.6

    - name: Install dependencies
      run: |
        python -m pip install --upgrade pip
        pip install flake8 pytest pytest-cov
        if [ -f requirements.txt ]; then pip install -r requirements.txt; fi

    - name: Install hazen
      run: |
        python setup.py install

    - name: test snr
      if: always()  # will always run regardless of whether previsu step fails - useful to ensure all CLI functions tested
      run: |
        hazen snr tests/data/Siemens --report

```

We would need to add a new step (the blocks that start with`- name`) for each CLI function. But I've not tested this, it's just to give a rough idea of what we need, please modify it so that it works! When you push to Github it will detect that this branch has an open PR and it will run the action. 

Note also that compared to the unit testing action i have removed the postgres db service since i don't think we'll need it here? Feel free to re-add it if required.

**Comment by pcw24601** [comment]
@hshuaib90 , @laurencejackson, I'm having trouble pushing this update--Sourcetree is giving me the following error--do I need extra permissions to write to this directory? Or is the problem at my end?

 ! [remote rejected] feature/HZN-104-tests-failing-to-run-on-develop- -> feature/HZN-104-tests-failing-to-run-on-develop- (refusing to allow an OAuth App to create or update workflow `.github/workflows/cli-test.yml` without `workflow` scope)
error: failed to push some refs to 'https://github.com/GSTT-CSC/hazen.git'

**Comment by pcw24601** [comment]
> @hshuaib90 , @laurencejackson, I'm having trouble pushing this update--Sourcetree is giving me the following error--do I need extra permissions to write to this directory? Or is the problem at my end?
> 
> ! [remote rejected] feature/HZN-104-tests-failing-to-run-on-develop- -> feature/HZN-104-tests-failing-to-run-on-develop- (refusing to allow an OAuth App to create or update workflow `.github/workflows/cli-test.yml` without `workflow` scope) error: failed to push some refs to 'https://github.com/GSTT-CSC/hazen.git'

Weird--looks like I can create the file on github.com website , but can push from Sourcetree. Anyhow, hopefully it's working now.

**Comment by pcw24601** [comment]
Hi @laurencejackson, this is ready for review. Thanks for your helpful comments--hopefully I've implemented them as you intended.

**Review by laurencejackson — APPROVED** [review]
@pcw24601 looks great! Happy for now to just test SNR and approve the PR but in the future we should expand the scope of this to test all the Hazen tasks. I'll open a new issue for this (#133)

Thanks for this!
